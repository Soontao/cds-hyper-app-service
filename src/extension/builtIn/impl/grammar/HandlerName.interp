token literal names:
null
null
null
'_'
null
null
null
null
null
null
null
null
null
null
null
null
null
null
null
null
null
null
null
null
null

token symbolic names:
null
WS
LINE_BREAK
UNDER_SCORE
UPPER_CHAR
CHAR
ACTION
FUNCTION
EVENT
CREATE
UPDATE
DELETE
READ
POST
GET
PATCH
NEW
EDIT
SAVE
CANCEL
BEFORE
ON
AFTER
FOR
EACH

rule names:
name
eventHandler
freedomEventHandler
actionHandler
entityName
actionName
eventName
each
freeName
event
hook
keywords


atn:
[4, 1, 24, 87, 2, 0, 7, 0, 2, 1, 7, 1, 2, 2, 7, 2, 2, 3, 7, 3, 2, 4, 7, 4, 2, 5, 7, 5, 2, 6, 7, 6, 2, 7, 7, 7, 2, 8, 7, 8, 2, 9, 7, 9, 2, 10, 7, 10, 2, 11, 7, 11, 1, 0, 1, 0, 1, 0, 3, 0, 28, 8, 0, 1, 0, 3, 0, 31, 8, 0, 1, 0, 3, 0, 34, 8, 0, 1, 1, 1, 1, 3, 1, 38, 8, 1, 1, 1, 3, 1, 41, 8, 1, 1, 1, 1, 1, 3, 1, 45, 8, 1, 1, 2, 1, 2, 1, 2, 1, 2, 1, 3, 1, 3, 3, 3, 53, 8, 3, 1, 3, 1, 3, 1, 3, 1, 3, 3, 3, 59, 8, 3, 1, 4, 1, 4, 1, 5, 1, 5, 1, 6, 1, 6, 1, 7, 1, 7, 1, 8, 3, 8, 70, 8, 8, 1, 8, 1, 8, 1, 8, 1, 8, 1, 8, 4, 8, 77, 8, 8, 11, 8, 12, 8, 78, 1, 9, 1, 9, 1, 10, 1, 10, 1, 11, 1, 11, 1, 11, 0, 0, 12, 0, 2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 22, 0, 4, 1, 0, 6, 7, 1, 0, 9, 19, 1, 0, 20, 22, 2, 0, 6, 8, 23, 24, 89, 0, 27, 1, 0, 0, 0, 2, 35, 1, 0, 0, 0, 4, 46, 1, 0, 0, 0, 6, 50, 1, 0, 0, 0, 8, 60, 1, 0, 0, 0, 10, 62, 1, 0, 0, 0, 12, 64, 1, 0, 0, 0, 14, 66, 1, 0, 0, 0, 16, 69, 1, 0, 0, 0, 18, 80, 1, 0, 0, 0, 20, 82, 1, 0, 0, 0, 22, 84, 1, 0, 0, 0, 24, 28, 3, 6, 3, 0, 25, 28, 3, 2, 1, 0, 26, 28, 3, 4, 2, 0, 27, 24, 1, 0, 0, 0, 27, 25, 1, 0, 0, 0, 27, 26, 1, 0, 0, 0, 28, 30, 1, 0, 0, 0, 29, 31, 5, 2, 0, 0, 30, 29, 1, 0, 0, 0, 30, 31, 1, 0, 0, 0, 31, 33, 1, 0, 0, 0, 32, 34, 5, 0, 0, 1, 33, 32, 1, 0, 0, 0, 33, 34, 1, 0, 0, 0, 34, 1, 1, 0, 0, 0, 35, 37, 3, 20, 10, 0, 36, 38, 3, 14, 7, 0, 37, 36, 1, 0, 0, 0, 37, 38, 1, 0, 0, 0, 38, 40, 1, 0, 0, 0, 39, 41, 3, 18, 9, 0, 40, 39, 1, 0, 0, 0, 40, 41, 1, 0, 0, 0, 41, 44, 1, 0, 0, 0, 42, 43, 5, 23, 0, 0, 43, 45, 3, 8, 4, 0, 44, 42, 1, 0, 0, 0, 44, 45, 1, 0, 0, 0, 45, 3, 1, 0, 0, 0, 46, 47, 3, 20, 10, 0, 47, 48, 5, 8, 0, 0, 48, 49, 3, 12, 6, 0, 49, 5, 1, 0, 0, 0, 50, 52, 3, 20, 10, 0, 51, 53, 3, 14, 7, 0, 52, 51, 1, 0, 0, 0, 52, 53, 1, 0, 0, 0, 53, 54, 1, 0, 0, 0, 54, 55, 7, 0, 0, 0, 55, 58, 3, 10, 5, 0, 56, 57, 5, 23, 0, 0, 57, 59, 3, 8, 4, 0, 58, 56, 1, 0, 0, 0, 58, 59, 1, 0, 0, 0, 59, 7, 1, 0, 0, 0, 60, 61, 3, 16, 8, 0, 61, 9, 1, 0, 0, 0, 62, 63, 3, 16, 8, 0, 63, 11, 1, 0, 0, 0, 64, 65, 3, 16, 8, 0, 65, 13, 1, 0, 0, 0, 66, 67, 5, 24, 0, 0, 67, 15, 1, 0, 0, 0, 68, 70, 5, 4, 0, 0, 69, 68, 1, 0, 0, 0, 69, 70, 1, 0, 0, 0, 70, 76, 1, 0, 0, 0, 71, 77, 5, 5, 0, 0, 72, 77, 3, 18, 9, 0, 73, 77, 3, 20, 10, 0, 74, 77, 3, 22, 11, 0, 75, 77, 3, 14, 7, 0, 76, 71, 1, 0, 0, 0, 76, 72, 1, 0, 0, 0, 76, 73, 1, 0, 0, 0, 76, 74, 1, 0, 0, 0, 76, 75, 1, 0, 0, 0, 77, 78, 1, 0, 0, 0, 78, 76, 1, 0, 0, 0, 78, 79, 1, 0, 0, 0, 79, 17, 1, 0, 0, 0, 80, 81, 7, 1, 0, 0, 81, 19, 1, 0, 0, 0, 82, 83, 7, 2, 0, 0, 83, 21, 1, 0, 0, 0, 84, 85, 7, 3, 0, 0, 85, 23, 1, 0, 0, 0, 11, 27, 30, 33, 37, 40, 44, 52, 58, 69, 76, 78]