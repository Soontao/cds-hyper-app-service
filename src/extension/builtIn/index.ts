export { HyperEntityHandler } from "./impl/HyperEntityHandler";
export { PageExample } from "./repo/PageExample";
export { BaseRepository } from "./repo/Repository";
