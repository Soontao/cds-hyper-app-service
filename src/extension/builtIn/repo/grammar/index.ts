/* eslint-disable @typescript-eslint/no-floating-promises */
/* eslint-disable max-len */
/* eslint-disable @typescript-eslint/ban-ts-comment */
import { cwdRequireCDS, EntityDefinition, fuzzy, memorized } from "cds-internal-tool";
import { DELETE, INSERT, SELECT, UPDATE } from "cds-internal-tool/lib/types/ql";
import antlr4 from "../../../base/vendor/antlr4";
import RepositoryLexer from "./RepositoryLexer";
import RepositoryListener from "./RepositoryListener";
import RepositoryParser from "./RepositoryParser";

type Query = SELECT<any> | UPDATE<any> | INSERT<any> | DELETE<any>

class RepositoryInformationListener extends RepositoryListener {

  private entity: EntityDefinition;

  private cds = cwdRequireCDS();

  private createQuery!: () => Query;

  // TODO: refactor this
  private params: Array<(query: Query, args: any) => void> = [];

  private argIndex = 0;

  private limit?: { one?: boolean, rows?: number, offset?: number };

  constructor(entity: EntityDefinition) {
    super();
    this.entity = entity;
  }

  private nextArgIndex(argNum = 1) {
    // get a temporary arg index
    const curArgIndex = this.argIndex;
    this.argIndex += argNum;
    return curArgIndex;
  }

  private toCQNLogic(logic: string, query: any) {
    // for the first parameter, must use the 'where' method
    return (
      query.SELECT?.where === undefined &&
      query.UPDATE?.where === undefined &&
      query.DELETE?.where === undefined
    ) ? "where" : logic;
  }

  private toCQNOp(op: string): string {
    switch (op) {
      case "EQUALS":
        return "=";
      case "NOTEQUALS":
        return "!=";
      case "IS":
        return "is";
      case "ISNOT":
        return "is not";
      case "LIKE":
        return "like";
      case "IN":
        return "in";
      default:
        break;
    }
    return "=";
  }

  private toLit(lit: string): any {
    switch (lit.toLowerCase()) {
      case "null":
        return null;
      case "true":
        return true;
      case "false":
        return false;
      default:
        break;
    }
  }


  enterFindQuery(): void {
    this.createQuery = () => {
      if (this.limit?.one === true) {
        return this.cds.ql.SELECT.one.from(this.entity);
      }
      return this.cds.ql.SELECT.from(this.entity);
    };
  }

  enterUpdateQuery(): void {
    this.createQuery = () => this.cds.ql.UPDATE.entity(this.entity);
    // first arguments is used to update
    const curArgIndex = this.nextArgIndex();
    this.params.push((query, args: Array<any>) => {
      (query as UPDATE<any>).with(args[curArgIndex]);
    });
  }

  enterDeleteQuery(): void {
    this.createQuery = () => this.cds.ql.DELETE.from(this.entity);
  }

  enterLimitExpr(ctx: any): void {
    if (ctx.ONE() !== null) {
      this.limit = { one: true };
    }
    if (ctx.topExpr()) {
      this.limit = {};
      this.limit.rows = parseInt(ctx.topExpr().NUMBER().getText(), 10);
      if (ctx.skipExpr()) {
        this.limit.offset = parseInt(ctx.skipExpr().NUMBER().getText(), 10);
      }
    }
  }

  enterFieldExpr(ctx: any): void {
    const fieldName = ctx.identifier().getText(); // TODO: map and check field is exist on entity or not
    const elementDef = fuzzy.findElement(this.entity, fieldName);
    if (elementDef === undefined) {
      throw new Error(`the field '${fieldName}' is not in entity '${this.entity.name}'`);
    }
    const elementName = elementDef.name;

    const op: string = String(ctx.operators()?.getText() ?? "EQUALS").toUpperCase();

    const lit = ctx.literal()?.getText();

    const rawLogic = (ctx.logic()?.getText?.() ?? "AND").toLowerCase();

    if (lit) {
      // plain literal
      this.params.push((query: Query) => {
        const litVal = this.toLit(lit);
        if (litVal === null) {
          query[this.toCQNLogic(rawLogic, query)](`${elementName} ${this.toCQNOp(op)} null`);
        } else {
          query[this.toCQNLogic(rawLogic, query)](`${elementName} ${this.toCQNOp(op)}`, this.toLit(lit));
        }
      });
    }
    else {
      const curArgIndex = this.nextArgIndex();
      this.params.push((query: Query, args: Array<any>) => {
        query[this.toCQNLogic(rawLogic, query)](`${elementName} ${this.toCQNOp(op)}`, args[curArgIndex]);
      });
    }
    // TODO: concern about two values op like BETWEEN
  }


  public toQuery(...args: Array<any>): Query {
    const query = this.createQuery();
    // TODO: debug param values
    this.params.forEach(param => param(query, args));
    if (this.limit !== undefined) {
      // @ts-ignore
      query.limit(this.limit.rows, this.limit.offset);
    }
    return query;
  }
}

class ThrowErrorListener extends antlr4.error.ErrorListener {
  syntaxError(_recognizer: any, _offendingSymbol: any, line: string, column: string, msg: string, _e: any) {
    throw new Error("Syntax Error: line " + line + ":" + column + " " + msg);
  }
}

/**
 * parser for methods in repository
 * 
 * @param entity 
 * @returns 
 */
export const createRepositoryParser = (entity: EntityDefinition) => memorized(function (repoMethodName: string) {
  const logger = cwdRequireCDS().log("cds-hyper-app-service");

  logger.debug("parse repository query", repoMethodName);

  try {
    const chars = new antlr4.InputStream(repoMethodName);
    const lexer = new RepositoryLexer(chars);
    const tokens = new antlr4.CommonTokenStream(lexer);
    // TODO: error messages
    lexer?.["removeErrorListeners"]?.();
    const parser = new RepositoryParser(tokens);
    parser?.["removeErrorListeners"]?.();
    parser?.["addErrorListener"](new ThrowErrorListener());
    parser["buildParseTrees"] = true;
    const listener = new RepositoryInformationListener(entity);
    const tree = parser.query();
    antlr4.tree.ParseTreeWalker.DEFAULT.walk(listener, tree);
    logger.debug("parse repository query succeed", repoMethodName);
    return (...args: Array<any>) => listener.toQuery(...args);
  } catch (error) {
    logger.debug("parse repository query", repoMethodName, "failed, not a valid repository query", error);
    throw error;
  }

});
