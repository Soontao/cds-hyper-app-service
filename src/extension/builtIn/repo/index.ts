export { RepositoryExt } from "./RepositoryExt";
export { isEmptyFunction } from "./utils";
