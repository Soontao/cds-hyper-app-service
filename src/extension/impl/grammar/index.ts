import { cwdRequireCDS, EventHook, EventName, memorized } from "cds-internal-tool";
import HandlerNameLexer from "./HandlerNameLexer";
import HandlerNameListener from "./HandlerNameListener";
import HandlerNameParser from "./HandlerNameParser";
import antlr4 from "./vendor/antlr4";


type HandlerNameInformation = {
  hooks: Array<EventHook>;
  events: Array<EventName>
  entity?: string;
  valid: boolean;
}

class HandlerNameInformationListener extends HandlerNameListener {

  private entity?: string;

  private events: Array<EventName> = [];

  private hooks: Array<EventHook> = [];

  exitHook(ctx: any): void {
    // only direct level hook
    if (ctx.parentCtx instanceof HandlerNameParser.NameContext) {
      const hook = ctx.getText();
      if (hook !== undefined) {
        this.hooks.push(String(hook).toLowerCase() as any);
      }
    }
  }

  exitEvent(ctx: any): void {
    const event = ctx.getText();
    if (event !== undefined) {
      this.events.push(String(event).toUpperCase() as any);
    }
  }

  exitEntity(ctx: any): void {
    const entity = ctx.getText();
    if (entity !== undefined) {
      this.entity = entity;
    }
  }

  /**
   * get the final information
   */
  public information(): HandlerNameInformation {
    const info: HandlerNameInformation = { hooks: this.hooks, events: this.events, valid: true };
    if (this.entity !== undefined) {
      info.entity = this.entity;
    }
    return info;
  }
}


/**
 * parse handler information from function name, return {valid:true} is name looks like a handler
 */
export const parseHandlerName = memorized(function (name: string) {
  try {
    const chars = new antlr4.InputStream(name);
    const lexer = new HandlerNameLexer(chars);
    const tokens = new antlr4.CommonTokenStream(lexer);
    const parser = new HandlerNameParser(tokens);
    parser["buildParseTrees"] = true;
    const listener = new HandlerNameInformationListener();
    const tree = parser.name();
    antlr4.tree.ParseTreeWalker.DEFAULT.walk(listener, tree);
    return listener.information();
  } catch (error) {
    cwdRequireCDS().log("cds-hyper-app-service").debug("parse name", name, "failed, not a valid handler name");
    return { valid: false, events: [], hooks: [] };
  }

});
