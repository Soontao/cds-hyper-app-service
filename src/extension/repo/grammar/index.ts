/* eslint-disable max-len */
/* eslint-disable @typescript-eslint/ban-ts-comment */
import { cwdRequireCDS, EntityDefinition, fuzzy, memorized } from "cds-internal-tool";
import { DELETE, INSERT, SELECT, UPDATE } from "cds-internal-tool/lib/types/ql";
import antlr4 from "../../base/vendor/antlr4";
import RepositoryLexer from "./RepositoryLexer";
import RepositoryListener from "./RepositoryListener";
import RepositoryParser from "./RepositoryParser";

type Query = SELECT<any> | UPDATE<any> | INSERT<any> | DELETE<any>

class RepositoryInformationListener extends RepositoryListener {

  private entity: EntityDefinition;

  private cds = cwdRequireCDS();

  private createQuery!: () => Query;

  private params: Array<(query: Query, args: any) => void> = [];

  private argIndex = 0;

  private limit?: { one?: boolean, rows?: number, offset?: number };

  constructor(entity: EntityDefinition) {
    super();
    this.entity = entity;
  }

  private nextArgIndex(argNum = 1) {
    // get a temporary arg index
    const curArgIndex = this.argIndex;
    this.argIndex += argNum;
    return curArgIndex;
  }

  private toCQNLogic(logic: string, argIndex: number) {
    // for the first parameter, must use the 'where' method
    return argIndex === 0 ? "where" : logic;
  }

  private toCQNOp(op: string): string {
    switch (op) {
      case "EQUALS": case "IS":
        return "=";
      case "LIKE":
        return "like";
      default:
        break;
    }
    return "=";
  }

  private toLit(lit: string): any {
    switch (lit.toLowerCase()) {
      case "null":
        return null;
      case "true":
        return true;
      case "false":
        return false;
      default:
        break;
    }
  }


  enterFindQuery(): void {
    this.createQuery = () => {
      if (this.limit?.one === true) {
        return this.cds.ql.SELECT.one.from(this.entity);
      }
      return this.cds.ql.SELECT.from(this.entity);
    };
  }

  enterUpdateQuery(): void {
    this.createQuery = () => this.cds.ql.UPDATE.entity(this.entity);
  }

  enterDeleteQuery(): void {
    this.createQuery = () => this.cds.ql.DELETE.from(this.entity);
  }

  enterLimitExpr(ctx: any): void {
    if (ctx.ONE() !== null) {
      this.limit = { one: true };
    }
    if (ctx.topExpr()) {
      this.limit = {};
      this.limit.rows = parseInt(ctx.topExpr().NUMBER().getText(), 10);
      if (ctx.skipExpr()) {
        this.limit.offset = parseInt(ctx.skipExpr().NUMBER().getText(), 10);
      }
    }
  }

  enterFieldExpr(ctx: any): void {
    const fieldName = ctx.identifier().getText(); // TODO: map and check field is exist on entity or not
    const elementDef = fuzzy.findElement(this.entity, fieldName);
    if (elementDef === undefined) {
      throw new Error(`the field '${fieldName}' is not in entity ${this.entity.name}`);
    }
    const elementName = elementDef.name;

    const operators: Array<string> = ctx.operators().length === 0 ?
      ["EQUALS"] :
      ctx.operators().map((op: any) => op.getText().toUpperCase());

    const lit = ctx.literal()?.getText();

    const rawLogic = (ctx.logic()?.getText?.() ?? "AND").toLowerCase();
    if (operators.length === 1) {
      const [op] = operators;
      if (op === "EQUALS" || op === "IS" || op === "LIKE") {
        if (lit) {
          this.params.push((query: Query) => {
            query[this.toCQNLogic(rawLogic, this.argIndex)]({ [elementName]: { [this.toCQNOp(op)]: this.toLit(lit) } });
          });
        } else {
          const curArgIndex = this.nextArgIndex();
          this.params.push((query: Query, args: Array<any>) => {
            query[this.toCQNLogic(rawLogic, curArgIndex)]({ [elementName]: { [this.toCQNOp(op)]: args[curArgIndex] } });
          });
        }

      }
    }
    // TODO: concern about two values op like BETWEEN
  }


  public toQuery(...args: Array<any>): Query {
    const query = this.createQuery();
    // TODO: debug param values
    this.params.forEach(param => param(query, args));
    if (this.limit !== undefined) {
      // @ts-ignore
      query.limit(this.limit.rows, this.limit.offset);
    }
    return query;
  }
}

class ThrowErrorListener extends antlr4.error.ErrorListener {
  syntaxError(_recognizer: any, _offendingSymbol: any, line: string, column: string, msg: string, _e: any) {
    throw new Error("Syntax Error: line " + line + ":" + column + " " + msg);
  }
}

export const createRepositoryParser = (entity: EntityDefinition) => memorized(function (repoMethodName: string) {
  const logger = cwdRequireCDS().log("cds-hyper-app-service");

  logger.debug("parse repository query", repoMethodName);

  try {
    const chars = new antlr4.InputStream(repoMethodName);
    const lexer = new RepositoryLexer(chars);
    const tokens = new antlr4.CommonTokenStream(lexer);
    // TODO: error messages
    lexer?.["removeErrorListeners"]?.();
    const parser = new RepositoryParser(tokens);
    parser?.["removeErrorListeners"]?.();
    parser?.["addErrorListener"](new ThrowErrorListener());
    parser["buildParseTrees"] = true;
    const listener = new RepositoryInformationListener(entity);
    const tree = parser.query();
    antlr4.tree.ParseTreeWalker.DEFAULT.walk(listener, tree);
    logger.debug("parse repository query succeed", repoMethodName);
    return (...args: Array<any>) => listener.toQuery(...args);
  } catch (error) {
    logger.debug("parse repository query", repoMethodName, "failed, not a valid repository query", error);
    throw error;
  }

});
