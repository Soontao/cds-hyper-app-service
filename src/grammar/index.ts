// @ts-nocheck
import { EventHook, EventName, memorized } from 'cds-internal-tool';
import HandlerNameLexer from './HandlerNameLexer';
import HandlerNameListener from "./HandlerNameListener";
import HandlerNameParser from './HandlerNameParser';
import antlr4 from './vendor/antlr4';


interface HandlerNameInformation {
  hooks: Array<EventHook>;
  events: Array<EventName>
  entity?: string;
}

class HandlerNameInformationListener extends HandlerNameListener {

  private entity?: string;
  private events: Array<EventName> = [];
  private hooks: Array<EventHook> = []

  exitHook(ctx: any): void {
    // only direct level hook
    if (ctx.parentCtx instanceof HandlerNameParser.NameContext) {
      const hook = ctx.getText()
      if (hook !== undefined) {
        this.hooks.push(String(hook).toLowerCase())
      }
    }
  }

  exitEvent(ctx: any): void {
    const event = ctx.getText()
    if (event !== undefined) {
      this.events.push(String(event).toUpperCase())
    }
  }

  exitEntity(ctx: any): void {
    const entity = ctx.getText()
    if (entity !== undefined) {
      this.entity = entity
    }
  }

  /**
   * get the final information
   */
  public information(): HandlerNameInformation {
    const info: HandlerNameInformation = { hooks: this.hooks, events: this.events }
    if (this.entity !== undefined) {
      info.entity = this.entity
    }
    return info
  }
}


export const parseHandlerName = memorized(function parseName(name: string) {
  const chars = new antlr4.InputStream(name);
  const lexer = new HandlerNameLexer(chars);
  const tokens = new antlr4.CommonTokenStream(lexer);
  const parser = new HandlerNameParser(tokens);
  parser['buildParseTrees'] = true;
  const listener = new HandlerNameInformationListener()
  const tree = parser.name()
  antlr4.tree.ParseTreeWalker.DEFAULT.walk(listener, tree)
  return listener.information()
})
