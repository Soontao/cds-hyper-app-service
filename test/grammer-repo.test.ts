// @ts-nocheck
import { cwdRequireCDS } from "cds-internal-tool";
import { isEmptyFunction } from "../src/extension/builtIn/repo";
import { createRepositoryParser } from "../src/extension/builtIn/repo/grammar";


describe("Repository Grammar Test Suite", () => {

  const cds = cwdRequireCDS()

  const { SELECT, UPDATE, DELETE } = cds.ql

  const model = cds.reflect(cds.parse.cdl('entity Demo {Name:String(255);age:Integer64;active:Boolean; }'))
  const demoEntityDef = model.definitions.Demo
  const parse = createRepositoryParser(demoEntityDef)

  it('should support parse simple method', () => {
    for (const methodName of ['findByName', 'findByNameEquals']) {
      const query = parse(methodName)('Theo')
      expect(query).toMatchSnapshot()
    }
  });

  it('should support parse multi field', () => {
    for (const methodName of ['findByNameAndAge', 'findByNameEqualsAndAgeEquals']) {
      const query = parse(methodName)('Theo', 12)
      expect(query).toMatchSnapshot()
    }
  });

  it('should support parse or keyword', () => {
    for (const methodName of ['findByNameOrAge', 'findByNameEqualsOrAgeEquals']) {
      const query = parse(methodName)('Theo', 12)
      expect(query).toMatchSnapshot()
    }
  });

  it('should support find one', () => {
    expect(parse('findOneByName')('Theo'))
      .toMatchSnapshot()
  });


  it('should support find one not equals', () => {
    expect(parse('findOneByNameNotEquals')('Theo'))
      .toMatchSnapshot()
  });

  it('should support find top', () => {
    expect(parse('findTop5ByName')('Theo'))
      .toMatchSnapshot()
  });

  it('should support find top skip', () => {
    expect(parse('findTop5Skip20ByName')('Theo'))
      .toMatchSnapshot()
  });

  it('should support like operator', () => {
    for (const methodName of ['findByNameLike']) {
      const query = parse(methodName)('Theo')
      expect(query).toMatchSnapshot()
    }
  });

  it('should support other operations', () => {
    const query = parse('deleteByName')("Theo")
    expect(query).toMatchSnapshot()
  });

  it('should support literal fixed value', () => {
    expect(parse("findByActiveEqualsTrue")()).toStrictEqual(SELECT.from("Demo").where({ active: true }))
    expect(parse("findByActiveEqualsFalse")()).toStrictEqual(SELECT.from("Demo").where({ active: false }))
    expect(parse("findByActiveEqualsNull")()).toStrictEqual(SELECT.from("Demo").where({ active: null }))
  });

  it('should support literal fixed value (not)', () => {
    expect(parse("findByActiveNotEqualsTrue")()).toStrictEqual(SELECT.from("Demo").where({ active: { ['!=']: true } }))
    expect(parse("findByActiveNotEqualsFalse")()).toStrictEqual(SELECT.from("Demo").where({ active: { ['!=']: false } }))
    expect(parse("findByActiveNotEqualsNull")()).toStrictEqual(SELECT.from("Demo").where({ active: { ['!=']: null } }))
  });

  it('should support is null', () => {
    expect(parse("findByActiveIsNull")()).toStrictEqual(SELECT.from("Demo").where`active is null`)
    expect(parse("findByActiveIsNotNull")()).toStrictEqual(SELECT.from("Demo").where`active is not null`)
  });


  it('should support "update" method', () => {
    expect(parse("updateByNameAge")({ active: true }, 'Theo', 19))
      .toStrictEqual(UPDATE.entity("Demo").set({ active: true }).where({ Name: 'Theo', age: 19 }))
  });


  it('should support raise error', () => {
    expect(() => parse('reuse-people-q-name')).toThrow("Syntax Error: ")
  });

  it('should raise error when not have column', () => {
    expect(() => parse('findByHeight')).toThrow(`the field 'Height' is not in entity 'Demo'`)
  });



  it('should support check function is empty or not', () => {
    function a() { }
    async function aa(params) { }
    async function aac(params) {
      // TODO: 
      // left empty for auto proxy
    }
    const b = () => { }
    const ba = async () => { }
    const bac = async () => {
      // left empty for auto proxy
    }

    const ci = new class {
      async cam() {
        // TODO: update later
      }
      cm() {
        // with comments
      }
      async cb(a, b) {
        return a + b
      }
    }

    for (const f of [a, aa, aac, b, ba, bac, ci.cam, ci.cm]) {
      expect(isEmptyFunction(f)).toBeTruthy()
    }

    for (const f of [ci.cb]) {
      expect(isEmptyFunction(f)).toBeFalsy()
    }

  });



});
