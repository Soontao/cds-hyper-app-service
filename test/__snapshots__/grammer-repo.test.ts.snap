// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Repository Grammar Test Suite should support find one 1`] = `
Query {
  "SELECT": {
    "from": {
      "ref": [
        "Demo",
      ],
    },
    "one": true,
    "where": [
      {
        "ref": [
          "Name",
        ],
      },
      "=",
      {
        "val": "Theo",
      },
    ],
  },
}
`;

exports[`Repository Grammar Test Suite should support find one not equals 1`] = `
Query {
  "SELECT": {
    "from": {
      "ref": [
        "Demo",
      ],
    },
    "one": true,
    "where": [
      {
        "ref": [
          "Name",
        ],
      },
      "!=",
      {
        "val": "Theo",
      },
    ],
  },
}
`;

exports[`Repository Grammar Test Suite should support find top 1`] = `
Query {
  "SELECT": {
    "from": {
      "ref": [
        "Demo",
      ],
    },
    "limit": {
      "rows": {
        "val": 5,
      },
    },
    "where": [
      {
        "ref": [
          "Name",
        ],
      },
      "=",
      {
        "val": "Theo",
      },
    ],
  },
}
`;

exports[`Repository Grammar Test Suite should support find top skip 1`] = `
Query {
  "SELECT": {
    "from": {
      "ref": [
        "Demo",
      ],
    },
    "limit": {
      "offset": {
        "val": 20,
      },
      "rows": {
        "val": 5,
      },
    },
    "where": [
      {
        "ref": [
          "Name",
        ],
      },
      "=",
      {
        "val": "Theo",
      },
    ],
  },
}
`;

exports[`Repository Grammar Test Suite should support like operator 1`] = `
Query {
  "SELECT": {
    "from": {
      "ref": [
        "Demo",
      ],
    },
    "where": [
      {
        "ref": [
          "Name",
        ],
      },
      "like",
      {
        "val": "Theo",
      },
    ],
  },
}
`;

exports[`Repository Grammar Test Suite should support other operations 1`] = `
Query {
  "DELETE": {
    "from": "Demo",
    "where": [
      {
        "ref": [
          "Name",
        ],
      },
      "=",
      {
        "val": "Theo",
      },
    ],
  },
}
`;

exports[`Repository Grammar Test Suite should support parse multi field 1`] = `
Query {
  "SELECT": {
    "from": {
      "ref": [
        "Demo",
      ],
    },
    "where": [
      {
        "ref": [
          "Name",
        ],
      },
      "=",
      {
        "val": "Theo",
      },
      "and",
      {
        "ref": [
          "age",
        ],
      },
      "=",
      {
        "val": 12,
      },
    ],
  },
}
`;

exports[`Repository Grammar Test Suite should support parse multi field 2`] = `
Query {
  "SELECT": {
    "from": {
      "ref": [
        "Demo",
      ],
    },
    "where": [
      {
        "ref": [
          "Name",
        ],
      },
      "=",
      {
        "val": "Theo",
      },
      "and",
      {
        "ref": [
          "age",
        ],
      },
      "=",
      {
        "val": 12,
      },
    ],
  },
}
`;

exports[`Repository Grammar Test Suite should support parse or keyword 1`] = `
Query {
  "SELECT": {
    "from": {
      "ref": [
        "Demo",
      ],
    },
    "where": [
      {
        "ref": [
          "Name",
        ],
      },
      "=",
      {
        "val": "Theo",
      },
      "or",
      {
        "ref": [
          "age",
        ],
      },
      "=",
      {
        "val": 12,
      },
    ],
  },
}
`;

exports[`Repository Grammar Test Suite should support parse or keyword 2`] = `
Query {
  "SELECT": {
    "from": {
      "ref": [
        "Demo",
      ],
    },
    "where": [
      {
        "ref": [
          "Name",
        ],
      },
      "=",
      {
        "val": "Theo",
      },
      "or",
      {
        "ref": [
          "age",
        ],
      },
      "=",
      {
        "val": 12,
      },
    ],
  },
}
`;

exports[`Repository Grammar Test Suite should support parse simple method 1`] = `
Query {
  "SELECT": {
    "from": {
      "ref": [
        "Demo",
      ],
    },
    "where": [
      {
        "ref": [
          "Name",
        ],
      },
      "=",
      {
        "val": "Theo",
      },
    ],
  },
}
`;

exports[`Repository Grammar Test Suite should support parse simple method 2`] = `
Query {
  "SELECT": {
    "from": {
      "ref": [
        "Demo",
      ],
    },
    "where": [
      {
        "ref": [
          "Name",
        ],
      },
      "=",
      {
        "val": "Theo",
      },
    ],
  },
}
`;
